src/sinks/aws_cloudwatch_logs/mod.rs

```diff
@@ -290,72 +281,48 @@ impl CloudwatchLogsSvc {
 impl Service<Vec<Event>> for CloudwatchLogsSvc {
     type Response = ();
     type Error = CloudwatchError;
-    type Future = Box<dyn Future<Item = Self::Response, Error = Self::Error> + Send + 'static>;
+    type Future = request::CloudwatchFuture;
 
     fn poll_ready(&mut self) -> Poll<(), Self::Error> {
-        loop {
-            match &mut self.state {
-                State::Idle => {
-                    let fut = self.describe_stream();
-                    self.state = State::Describe(fut);
-                    continue;
-                }
-                State::Describe(fut) => {
-                    let response = try_ready!(fut.poll().map_err(CloudwatchError::Describe));
-
-                    let stream = if let Some(stream) = response
-                        .log_streams
-                        .ok_or(CloudwatchError::NoStreamsFound)?
-                        .into_iter()
-                        .next()
-                    {
-                        stream
-                    } else {
-                        let fut = self.create_log_stream();
-                        self.state = State::CreateStream(fut);
-                        continue;
-                    };
-
-                    self.state = State::Token(stream.upload_sequence_token);
-                    return Ok(Async::Ready(()));
-                }
-                State::CreateStream(fut) => {
-                    let _ = try_ready!(fut.poll().map_err(CloudwatchError::CreateStream));
-                    self.state = State::Idle;
-                    continue;
+        if self.in_flight {
+            match self.token_rx.poll() {
+                Ok(Async::Ready(Some(token))) => {
+                    self.in_flight = false;
+                    self.token = token;
+                    return Ok(().into());
                 }
-                State::Token(_) => return Ok(Async::Ready(())),
-                State::Put(fut) => {
-                    let response = match fut.poll() {
-                        Ok(Async::Ready(response)) => response,
-                        Ok(Async::NotReady) => return Ok(Async::NotReady),
-                        Err(_) => panic!("The in flight future was dropped!"),
-                    };
-
-                    self.state = State::Token(response.next_sequence_token);
-                    return Ok(Async::Ready(()));
+                Ok(Async::Ready(None)) => {
+                    // Since we own one of the `tx` there will
+                    // always be one active one thus we will never
+                    // hit this none case until CloudwatchLogSvc gets dropped.
+                    unreachable!()
```

Comment:
I wonder if we could get rid of this by replacing the channel with something like a oneshot and making this service more of an explicit state machine.